### PR TITLE
Rek shipment endpoints

### PR DESCRIPTION
**Pull request by kilbergr**
This PR is part of the TSP epic. It includes api endpoint for shipments, querying the database for the index of created shipments and filtering by type of shipment (awarded or available). Corresponds to ticket https://www.pivotaltracker.com/story/show/154932671 in pivotal.

### CODE REVIEW

File: pkg/handlers/shipments.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	shipmentop "github.com/transcom/mymove/pkg/gen/restapi/operations/shipments"
+	"github.com/transcom/mymove/pkg/models"
+)
+
+type possiblyAwardedShipment struct {
+	ID                              uuid.UUID  `db:"id"`
+	CreatedAt                       time.Time  `db:"created_at"`
+	UpdatedAt                       time.Time  `db:"updated_at"`
+	TrafficDistributionListID       uuid.UUID  `db:"traffic_distribution_list_id"`
+	TransportationServiceProviderID nulls.UUID `db:"transportation_service_provider_id"`
```
Comment (macrael) [inline pkg/handlers/shipments.go]: I know that Pop uses the nulls package to represent nullability, but for ease of translation with go-swagger which is using pointers to represent nullability, I think we should use the same here. We've been doing that in our other models and haven't had any issues. Please let me know if pointers don't work and we can switch to using the nulls package everywhere. 

I think that if you set the type to *uuid.UUID instead of nulls.UUID the DB query should still work as expected.

File: pkg/handlers/shipments.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+// IndexShipmentsHandler returns a list of all shipments
+func IndexShipmentsHandler(p shipmentop.IndexShipmentsParams) middleware.Responder {
+	var response middleware.Responder
+
+	shipments := []possiblyAwardedShipment{}
+
+	// TODO Can Q() be .All(&shipments)
+	query := dbConnection.Q().LeftJoin("awarded_shipments", "awarded_shipments.shipment_id=shipments.id")
+
+	sql, args := query.ToSQL(&pop.Model{Value: models.Shipment{}},
```
Comment (macrael) [inline pkg/handlers/shipments.go]: As our SQL starts to get more complex, I think that we should start to think about moving all of our SQL into the models themselves. This is a bigger effort and needs to be applied broadly, but I think that it would be best if our handlers never had to think about SQL at all, and just called functions in the model packages to get what they need to operate on.

File: pkg/handlers/shipments.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+		"shipments.id",
+		"shipments.created_at",
+		"shipments.updated_at",
+		"shipments.traffic_distribution_list_id",
+		"awarded_shipments.transportation_service_provider_id",
+		"awarded_shipments.administrative_shipment",
+	)
+
```
Comment (macrael) [inline pkg/handlers/shipments.go]: As our SQL gets more complex, I think that we should be thinking about moving all of our SQL into the model file. This should be an effort everywhere, but I think that we should only actually talk to the database from inside our model classes. The handlers should just call functions to get what they need rather than needing to think about the SQL. This one presents an interesting problem, I think that it means for this API you'd want to move the possiblyAwardedShipment struct into the model file as well.

File: cmd/webserver/main.go [hunk 4]
```diff
@@ -76,9 +77,11 @@ func main() {
 
 	api.IssuesCreateIssueHandler = issueop.CreateIssueHandlerFunc(handlers.CreateIssueHandler)
 	api.IssuesIndexIssuesHandler = issueop.IndexIssuesHandlerFunc(handlers.IndexIssuesHandler)
+
 	api.Form1299sCreateForm1299Handler = form1299op.CreateForm1299HandlerFunc(handlers.CreateForm1299Handler)
 	api.Form1299sIndexForm1299sHandler = form1299op.IndexForm1299sHandlerFunc(handlers.IndexForm1299sHandler)
 
+	api.ShipmentsIndexShipmentsHandler = shipmentop.IndexShipmentsHandlerFunc(handlers.IndexShipmentsHandler)
```
Comment (macrael) [inline cmd/webserver/main.go]: Can these be ordered so they are grouped by API?

File: pkg/handlers/shipments.go (hunk 1 above)
Comment (jim) [reply pkg/handlers/shipments.go]: Makes sense! I will give this a shot.

File: pkg/handlers/shipments.go (hunk 3 above)
Comment (jim) [reply pkg/handlers/shipments.go]: 👍

Comment (jim) [reply pkg/handlers/shipments.go]: We actually were talking about this issue when we wrote it. We can move it into the model and see how that goes.

File: pkg/models/shipment.go [hunk 5]
```diff
@@ -2,24 +2,21 @@ package models
...
-	CreatedAt               time.Time `json:"created_at" db:"created_at"`
-	UpdatedAt               time.Time `json:"updated_at" db:"updated_at"`
-	TrafficDistributionList uuid.UUID `json:"traffic_distribution_list" db:"traffic_distribution_list"`
-}
-
-// String is not required by pop and may be deleted
-func (s Shipment) String() string {
-	js, _ := json.Marshal(s)
-	return string(js)
+	ID                        uuid.UUID `json:"id" db:"id"`
```
Comment (amitch23) [inline pkg/models/shipment.go]: This model is missing some values defined in the swagger.yaml and being handled in the handlers, ie. "name", "transportation_service_provider_id", and "administrative_shipment". Was this intentional?  Are the migrations for this model meant to be in another PR?

File: swagger.yaml [hunk 6]
```diff
@@ -761,3 +775,53 @@ definitions:
...
+        x-nullable: true
+      created_at:
+        type: string
+        format: date-time
+      updated_at:
+        type: string
+        format: date-time
+    required:
+      - id
+      - name
```
Comment (amitch23) [inline swagger.yaml]: You're setting "name" to required here but it doesn't seem to exist as a column in the shipment table or the shipment model - can you explain the reasoning behind this?

Comment (kilbergr) [reply swagger.yaml]: Ooh that's old will remove.

File: pkg/models/shipment.go [hunk 7]
```diff
@@ -18,10 +19,33 @@ type Shipment struct {
...
+	ID                              uuid.UUID  `db:"id"`
+	CreatedAt                       time.Time  `db:"created_at"`
+	UpdatedAt                       time.Time  `db:"updated_at"`
+	TrafficDistributionListID       uuid.UUID  `db:"traffic_distribution_list_id"`
+	TransportationServiceProviderID *uuid.UUID `db:"transportation_service_provider_id"`
+	AdministrativeShipment          *bool      `db:"administrative_shipment"`
+}
+
+// FetchPossiblyAwardedShipments runs the SQL query to fetch possibly awarded shipments from db
+func FetchPossiblyAwardedShipments(dbConnection *pop.Connection) ([]PossiblyAwardedShipment, error) {
```
Comment (akostibas) [inline pkg/models/shipment.go]: This seems fine for now, but I'm curious why we don't just go the rest of the way to find _actually_ awarded (and unawarded) shipments? Seems common enough that we'd want to save the caller the effort.

File: src/scenes/Shipments/Shipments.jsx [hunk 8]
```diff
@@ -10,17 +10,17 @@ import ShipmentCards from 'scenes/Shipments/ShipmentCards';
 import { loadShipments } from './ducks';
 
 export class Shipments extends Component {
+  componentDidMount() {
+    this.props.loadShipments();
```
Comment (akostibas) [inline src/scenes/Shipments/Shipments.jsx]: @kilbergr I assume y'all figured out how to make this reload the shipments when the URL changes?

Comment (jim) [reply src/scenes/Shipments/Shipments.jsx]: AFAIK that is an outstanding issue. Will check in with @tinyels in the morning for insight.

File: pkg/models/shipment.go (hunk 7 above)
Comment (jim) [reply pkg/models/shipment.go]: I think the rationale was that if all we're doing is piping data to the client and the filtering is done there, there wasn't a need for separate collections. I agree that we will probably end up needing that on the backend, and perhaps we even end up exposing both models to the client.

File: src/scenes/Shipments/Shipments.jsx (hunk 8 above)
Comment (kilbergr) [reply src/scenes/Shipments/Shipments.jsx]: This working has to do with filteredShipments being included in the component. So we're loading ALL shipments, whether awarded or not, and just filtering for display rather than only loading relevant shipments. This will undoubtedly change because we don't actually want to be loading all shipments every time.

Comment (tinyels) [reply src/scenes/Shipments/Shipments.jsx]: In general the component should not be using filter--we should create a selector for this. We can talk about what that means.
